src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -173,19 +170,17 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, Nu
                     refKind = rk;
                 }
 
-                if ((object)type.Type == NoReturnExpression)
-                {
-                    hasReturnWithoutArgument = true;
-                }
-                else
+                if ((object)type.Type != NoReturnExpression)
                 {
+                    hasReturnValue = true;
                     types.Add((returnStatement.ExpressionOpt, type));
                 }
             }
 
             HashSet<DiagnosticInfo> useSiteDiagnostics = null;
             var bestType = CalculateReturnType(compilation, conversions, delegateType, types, isAsync, node, ref useSiteDiagnostics);
-            return new InferredLambdaReturnType(types.Count, hasReturnWithoutArgument, refKind, bestType, useSiteDiagnostics.AsImmutableOrEmpty());
+            types.Free();
```

Comment:
>types.Free(); [](start = 12, length = 13)

Were we leaking this builder before?

Reply:
Yes.

---
In reply to: [410970357](https://github.com/dotnet/roslyn/pull/43366#discussion_r410970357) [](ancestors = 410970357)